Source/player.cpp

```diff
@@ -514,26 +514,44 @@ void StartSpell(int pnum, Direction d, int cx, int cy)
 		return;
 	}
 
+	// Checks conditions for spell again, cause initial check was done when spell was queued and the parameters could be changed meanwhile
+	bool isValid = true;
+	switch (player.queuedSpell.spellType) {
+	case RSPLTYPE_SKILL:
+	case RSPLTYPE_SPELL:
+		isValid = CheckSpell(MyPlayerId, player.queuedSpell.spellId, player.queuedSpell.spellType, true) == SpellCheckResult::Success;
+		break;
+	case RSPLTYPE_SCROLL:
+		isValid = UseScroll(player, player.queuedSpell.spellId);

```

Comment:
You are right. 🙂 That's why I wrote in the initial comment that we can have slightly more desync.
For reference #2681 happens much more often.
Another way to mitigate this would be to resolve #2724.

Personally I wouldn't set this as a show stopper.
The damage is still synced. The player location is updated frequently.
More important is that the player that generates the events can't cheat/abuse the bug anymore.

Reply:
The reason I consider this issue to be pretty significant is because it's guaranteed to affect all casting from scrolls when playing brand new characters, and it's also likely to affect nearly all casting from scrolls except Town Portal. The only argument I can think of to downplay this issue is the fact that most players probably don't use scrolls very often, if at all. But for players who do, I don't believe they'll typically be toting them around at the start of every game they play.

Reply:
Currently we need to decide between:
1. Double-cast bugs:
   I have only one scroll, but I can cast it twice.
   I have only enough mana for one cast, but I can cast twice.
   Affects gameplay / balance.
   Happens in single- **and** multiplayer.
   Reported in Issue #789.
   Present in current master.
3. **Visual** desync in multiplayer when a scroll is not synced or mana is not synced correctly
   The player A doesn't **see** Player B casting.
   Happens only in multiplayer.
   Resulting damage is still synced.
   Multiplayer still works.
   Will be introduced with this pr.

Ideally we shouldn't have any of these two bugs.
When #2861 is fixed this should be possible.
But if i have to decide I would go with the **visual** multiplayer-only bug.
That's why I think this pr is a improvement and would prefer to merge it. 🙂

Reply:
Why does that require that the game works on Mac from before 2010?

Reply:
Without old mac support everything is nothing. 😛 
Yeah sorry i meant #2724.

Reply:
Framing the fact that remote clients do not see the spell as being merely a visual desync is a bit overzealous. The spell gets cast on the local client, but does not get cast on the remote client. Try casting an slvl 15 Fire Wall from a scroll using this PR and let me know if you still think the issue is merely a visual desync.

Furthermore, there's nothing the player can do to get around this. If two new players are roaming through the dungeon together and they happen upon a Scroll of Chain Lightning, only one of them will actually be able to try out the spell. With this PR, the other will also inevitably not see what happened.

Let's also consider the implications of the double-casting bug. Sure, I can cast an Identify scroll twice in quick succession, but that doesn't mean I can use it to identify two items. I can cast an extra Fireball from a scroll, which means I can hit up to 8 or 9 monsters (but more likely 0-3 monsters) that I guess I shouldn't have been able to hit with Fireball. Maybe I can get two Stone Curses, occurring simultaneously, when I should have only been able to get one. These issues don't seem very impactful, and I'm grasping at straws. If there's a major implication of this bug I am not seeing that could justify fast-tracking this fix, please correct me.

I do not dispute that #2724 should help to resolve this issue, but I don't think that's an argument for merging this now. Instead, I think it makes more sense to sync player inventories first, then revisit this PR.

Reply:
> Framing the fact that remote clients do not see the spell as being merely a visual desync is a bit overzealous. The spell gets cast on the local client, but does not get cast on the remote client. Try casting an slvl 15 Fire Wall from a scroll using this PR and let me know if you still think the issue is merely a visual desync.

This following happens in the fire wall example:

- Player 1 cast a fire wall on monsters.
- Player 2 doesn't see the fire wall
- Monster dies on Player 1 client. Monster damage/kill gets synced.  (`CheckMissileCol` -> `MonsterMHit` -> `M_StartHit`/`M_StartKill`).
- Player 2 sees monster dying, cause of the synced damage/kill.

That means the monster still dies/take damage but you don't see it => that's why I concluded this is a only visual desync. But perhaps I miss some cases? 🤔

(Note if this is not only a visual desync #2681 is even more serious/critical then I thought, Cause there happens the same problem. Player 1 casts but Player 2 doesn't "see" it.)

> Furthermore, there's nothing the player can do to get around this. If two new players are roaming through the dungeon together and they happen upon a Scroll of Chain Lightning, only one of them will actually be able to try out the spell. With this PR, the other will also inevitably not see what happened.

You are correct Player 2 doesn't see the cast and this is a regression this pr would introdcue. But also in master only one player can "try out"/cast the spell or do I miss something?

> Let's also consider the implications of the double-casting bug. Sure, I can cast an Identify scroll twice in quick succession, but that doesn't mean I can use it to identify two items. I can cast an extra Fireball from a scroll, which means I can hit up to 8 or 9 monsters (but more likely 0-3 monsters) that I guess I shouldn't have been able to hit with Fireball. Maybe I can get two Stone Curses, occurring simultaneously, when I should have only been able to get one. These issues don't seem very impactful, and I'm grasping at straws. If there's a major implication of this bug I am not seeing that could justify fast-tracking this fix, please correct me.

For me it's some sort of cheating. You can do thing's you aren't allowed to do.

You can
- cast a scroll multiple times
- cast a spell/scroll/staff charge with higher spell level
- cast a spell without the needed mana

See the linked video in the description.

> I do not dispute that #2724 should help to resolve this issue, but I don't think that's an argument for merging this now. Instead, I think it makes more sense to sync player inventories first, then revisit this PR.

I think this should go in, even if #2724 isn't merged. That doesn't mean it needs to be in 1.4 release. I think this needs time, cause it changes some core mechanics and need well testing; even in singleplayer.

Perhaps we can work together to get this done. You are much more versed in regards to multiplayer then me. So perhaps you can work on #2724. Then we can get both in 1.5 release; a fixed bug and a nicely synchronized inventory. I would really appreciate that. 🙂

Reply:
> That means the monster still dies/take damage but you don't see it => that's why I concluded this is a only visual desync. But perhaps I miss some cases? 🤔

There's two important cases you missed, I think. First, monster AI will avoid flaming tiles. On Player 1's client, the monster may be stuck behind a Fire Wall, whereas on Player 2's client, the monster will walk right on through. In my testing, I found the monsters eventually just teleported to the other side of the Fire Wall on Player 1's client. Therefore, this issue can sabotage some early-game strategies, although admittedly early-game Fire Wall doesn't last very long.

Second, players can also get hurt by walking onto flaming tiles. In my testing, Player 1 thought Player 2 was taking damage whereas Player 2 thought he was standing on a normal tile. This situation can ultimately resolve itself okay, except that Player 2 probably should not have been able to just brazenly walk through the flaming tiles without taking damage. Also Player 1 won't be able to walk through even though Player 2 won't have any idea why.

> Perhaps we can work together to get this done. You are much more versed in regards to multiplayer then me. So perhaps you can work on #2724. Then we can get both in 1.5 release; a fixed bug and a nicely synchronized inventory. I would really appreciate that. 🙂

FYI, my comments here are just me sharing my thoughts after having reviewed this PR to determine if I think it should be included in 1.4. As you might imagine, I was thinking it should be probably be postponed so that these comments can be addressed. I certainly don't mind working on #2724 to see that this change makes it into 1.5.

Reply:
> You are correct Player 2 doesn't see the cast and this is a regression this pr would introdcue. But also in master only one player can "try out"/cast the spell or do I miss something?

I forgot to respond to this. You are correct that only one player can try out the spell, but both players should be able to see the result. A Barbarian playing with a Sorcerer, for instance, will likely be unable to use the scroll but would be able to see the effect if he watched the Sorcerer cast it. I wouldn't call it much of a shared experience for the two players otherwise.

Reply:
> There's two important cases you missed, I think. First, monster AI will avoid flaming tiles. On Player 1's client, the monster may be stuck behind a Fire Wall, whereas on Player 2's client, the monster will walk right on through. In my testing, I found the monsters eventually just teleported to the other side of the Fire Wall on Player 1's client. Therefore, this issue can sabotage some early-game strategies, although admittedly early-game Fire Wall doesn't last very long.
> 
> Second, players can also get hurt by walking onto flaming tiles. In my testing, Player 1 thought Player 2 was taking damage whereas Player 2 thought he was standing on a normal tile. This situation can ultimately resolve itself okay, except that Player 2 probably should not have been able to just brazenly walk through the flaming tiles without taking damage. Also Player 1 won't be able to walk through even though Player 2 won't have any idea why.

I think the most problematic spell is fire wall or? For the other spells it's mostly cosmetic/visual. 🤔 
And don't get me wrong, I would really like to get both issues (cheating/abuse & visual/sync) correct.

> FYI, my comments here are just me sharing my thoughts after having reviewed this PR to determine if I think it should be included in 1.4. As you might imagine, I was thinking it should be probably be postponed so that these comments can be addressed. I certainly don't mind working on #2724 to see that this change makes it into 1.5.

It would be a great if you can help to get #2724 forward. ❤️ 
And yes the 1.4 release is out of scope for this pr. Personally I would like to get this pr in early 1.5 version, so we can play test it a long time.